subsys/testsuite/ztest/src/ztest_new.c

```diff
@@ -0,0 +1,619 @@
+	switch (phase) {
+	case TEST_PHASE_SETUP:
+	case TEST_PHASE_BEFORE:
+	case TEST_PHASE_TEST:
+	case TEST_PHASE_AFTER:
+	case TEST_PHASE_TEARDOWN:
+		PRINT(" at %s function\n", phase_str[phase]);
+		longjmp(test_fail, 1);
+	case TEST_PHASE_FRAMEWORK:
+		PRINT("\n");
+		longjmp(stack_fail, 1);
+	}
+}
+
+static void init_testing(void)
+{
+	signal(SIGABRT, handle_signal);
+	signal(SIGSEGV, handle_signal);
+
+	if (setjmp(stack_fail)) {
+		PRINT("Test suite crashed.");
+		exit(1);
+	}
+}
+
+static int run_test(struct ztest_suite_node *suite, struct ztest_unit_test *test, void *data)
+{
+	int ret = TC_PASS;
+
+	TC_START(test->name);
+
+	if (setjmp(test_fail)) {
+		ret = TC_FAIL;
+		goto out;
+	}
+
+	if (setjmp(test_pass)) {
+		ret = TC_PASS;
+		goto out;
+	}
+
+	run_test_functions(suite, test, data);
+out:
+	ret |= cleanup_test(test);
+	Z_TC_END_RESULT(ret, test->name);
+
+	return ret;
+}
+
+#else /* KERNEL */
+
+/* Zephyr's probably going to cause all tests to fail if one test fails, so
+ * skip the rest of tests if one of them fails
+ */
+#ifdef CONFIG_ZTEST_FAIL_FAST
+#define FAIL_FAST 1
+#else
+#define FAIL_FAST 0
+#endif
+
+K_THREAD_STACK_DEFINE(ztest_thread_stack, CONFIG_ZTEST_STACKSIZE + CONFIG_TEST_EXTRA_STACKSIZE);
+static ZTEST_BMEM int test_result;
+
+static void test_finalize(void)
+{
+	if (IS_ENABLED(CONFIG_MULTITHREADING)) {
+		k_thread_abort(&ztest_thread);
+		k_thread_abort(k_current_get());
+	}
+}
+
+void ztest_test_fail(void)
+{
+	test_result = -1;
+	test_finalize();
+}
+
+void ztest_test_pass(void)
+{
+	test_result = 0;
+	test_finalize();
+}
+
+void ztest_test_skip(void)
+{
+	test_result = -2;
+	test_finalize();
+}
+
+void ztest_simple_1cpu_before(void *data)
+{
+	ARG_UNUSED(data);
+	z_test_1cpu_start();
+}
+
+void ztest_simple_1cpu_after(void *data)
+{
+	ARG_UNUSED(data);
+	z_test_1cpu_stop();
+}
+
+static void init_testing(void)
+{
+	k_object_access_all_grant(&ztest_thread);
+}
+
+static void test_cb(void *a, void *b, void *c)
+{
+	struct ztest_suite_node *suite = a;
+	struct ztest_unit_test *test = b;
+
+	test_result = 1;
+	run_test_functions(suite, test, c);
+	test_result = 0;
+}
+
+static int run_test(struct ztest_suite_node *suite, struct ztest_unit_test *test, void *data)
+{
+	int ret = TC_PASS;
+
+	TC_START(test->name);
+
+	if (IS_ENABLED(CONFIG_MULTITHREADING)) {
+		k_thread_create(&ztest_thread, ztest_thread_stack,
+				K_THREAD_STACK_SIZEOF(ztest_thread_stack),
+				(k_thread_entry_t)test_cb, suite, test, data,
+				CONFIG_ZTEST_THREAD_PRIORITY,
+				test->thread_options | K_INHERIT_PERMS, K_FOREVER);
+
+		if (test->name != NULL) {
+			k_thread_name_set(&ztest_thread, test->name);
+		}
+		k_thread_start(&ztest_thread);
+		k_thread_join(&ztest_thread, K_FOREVER);
+	} else {
+		test_result = 1;
+		run_test_functions(suite, test, data);
+	}
+
+	phase = TEST_PHASE_AFTER;
+	if (suite->after != NULL) {
+		suite->after(data);
+	}
+	run_test_rules(/*is_before=*/false, test, data);
+	phase = TEST_PHASE_FRAMEWORK;
+
```

Comment:
i recently added a feature to flush logs are each test case. Can you add it here as well? See [`ztest.c`](https://github.com/zephyrproject-rtos/zephyr/blob/521e6ffb7c1fcb663b98546eeba7822e03ca5a49/subsys/testsuite/ztest/src/ztest.c#L383)

Reply:
Done